Added option to import CFF files

AidanM11 opened the pull request:
<!-- 
Describe the changes you have made here: what, why, ... 
Link issues that are fixed, e.g. "Fixes #333".
If you fixed a koppor issue, link it, e.g. "Fixes https://github.com/koppor/jabref/issues/47".
The title of the PR must not reference an issue, because GitHub does not support autolinking there.
-->

Added ability to import CFF files as requested in #7945. As per CFF specifications, there is only one entry per CFF file, and CFF entries are always of the type "Software". JabRef does not support all the fields that CFF does, so a fair amount of information could be lost on import (specifically details about the authors beyond their names). I still need to test this more thoroughly and add automated tests, so I've marked this as a draft. I also plan on adding export support.

fixes #7945 

Images:
![Capture](https://user-images.githubusercontent.com/41391775/127573051-d78a949c-ea91-46ac-9bb9-48242ffbe9d7.PNG)
![Real2](https://user-images.githubusercontent.com/41391775/127573058-2d0f0c06-f443-491d-b31a-d03fb7ad5ece.PNG)
![Capture3](https://user-images.githubusercontent.com/41391775/127573065-eaa70449-6c87-4ae2-970a-c5b63bd2c8cd.PNG)


<!-- 
- Go through the list below. Please don't remove any items.
- [x] done; [ ] not done / not applicable
-->

- [x] Change in `CHANGELOG.md` described in a way that is understandable for the average user (if applicable)
- [x] Tests created for changes (if applicable)
- [x] Manually tested changed features in running JabRef (always required)
- [x] Screenshots added in PR description (for UI changes)
- [x] [Checked documentation](https://docs.jabref.org/): Is the information available and up to date? If not created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, submitted a pull request to the documentation repository.

Siedlerchr commented:
Thanks for tackling this issue. For the Authors take a look at Jabrefs Author List and Authors parser. Maybe you can use that.
For fields that do not map to standard fields, you could import/add them with a prefix

Siedlerchr commented:
For the java classes from the yaml, you can generate them from https://www.jsonschema2pojo.org/

Siedlerchr commented:
@AidanM11  What is the status here? It would be really cool if you could continue working on this

AidanM11 commented:
Sorry about the delay, had a busy couple weeks. I've added tests and made some minor changes to the code. As of now, the importer can parse and map most of the fields CFF supports. The main exception is the person/entity system from CFF. In CFF, the authors and contact fields are lists of Person or Entity objects, which each have their own extensive lists of fields for all kinds of information. It seems like it would be very difficult to fit that system into the way jabref handles fields, so instead I set it up so that the importer just extracts the names of each object in the 'authors' list and concatenates them into an author string. In this case, the import operation seems unavoidably lossy. 

Similarly, CFF supports a list of 'identifiers' of different types (url, swh, other). JabRef does have support for these fields, but the CFF specs appear allow the user to add as many of these identifiers as they like (I'm not entirely sure about this, but it seems like it). So a single entry could have multiple SWH identifiers, which seems tricky to work into the JabRef system.

However, there are also some 'static' fields in CFF that don't have JabRef counterparts (specifically: commit, license-url, repository-code, repository-artifact). You mentioned earlier that if CFF fields don't map to JabRef standard fields, I could "import/add them with a prefix". I wasn't sure what that meant, and was unable to find any examples of it in the codebase, so I was wondering if I could get a little more clarification on that.

So, overall, the importer is nearly finished, I just need to work out what to do with identifiers (maybe just use the first one, or do a comma-separated list?) and also what to do about the 'static' fields without mappings.

Siedlerchr commented on src/main/java/org/jabref/logic/importer/fileformat/CffImporter.java:
```diff
@@ -0,0 +1,155 @@
+        hm.put("abstract", StandardField.ABSTRACT);
+        hm.put("message", StandardField.COMMENT);
+        hm.put("date-released", StandardField.DATE);
+
```
For fields that can't be mapped, you can simply create a new UnkownField: (example is taken from the MSBibImporter)
And set CFF_PREFIX to sth ilke `cff-`
`new UnknownField(CFF_PREFIX + "medium"), "Medium"); `

JabRef can show non-standard fields in the entry editor as well

tobiasdiez replied:
I think there should be a better solution than prefixing those fields, since none of the citation styles would pick them up - but you might want to have the repository printed in the reference list. I've started a discussion at the biblatex repo how to handle those fields: https://github.com/plk/biblatex/issues/1169

Siedlerchr commented on src/main/java/org/jabref/logic/importer/fileformat/CffImporter.java:
```diff
@@ -0,0 +1,155 @@
+            String aAlias = auth.vals.get("alias");
+
+            if (aName != null) {
+                authors.add(aName);
```
JabRef has a class called Author which will handle the construction of authors, see org.jabref.model.entry
`    public Author(String first, String firstabbr, String von, String last, String jr) 
`
Afterwards, you put in an AuthorLIst  and return the and separated value
See for example the Medra-Fetcher  (org.jabref.logic.importer.fetcher)

sdruskat replied:
Apart from `String firstabbr`, CFF has the same fields for `person` type objects (e.g., authors). The main difference between these and the `entity` type objects (for this use case at least) is that entities only have a single field `name` instead, which I guess could be imported as is?

jspaaks replied:
- `first` -> [`given-names`](https://github.com/citation-file-format/citation-file-format/blob/1.2.0/schema-guide.md#definitionspersongiven-names)
- `von` -> [`name-particle`](https://github.com/citation-file-format/citation-file-format/blob/1.2.0/schema-guide.md#definitionspersonname-particle)
- `last` -> [`family-names`](https://github.com/citation-file-format/citation-file-format/blob/1.2.0/schema-guide.md#definitionspersonfamily-names)
- `jr` -> [`name-suffix`](https://github.com/citation-file-format/citation-file-format/blob/1.2.0/schema-guide.md#definitionspersonname-suffix)

Siedlerchr replied:
Yep, that seems to be correct, for reference:
https://github.com/JabRef/jabref/blob/cef4151c3805b080bf2a11b71d3182fdb5fd2446/src/main/java/org/jabref/model/entry/Author.java#L22-L31

sdruskat commented:
Hi :wave:, CFF co-lead here. It's really great to see this work happening for JabRef!

> As per CFF specifications, there is only one entry per CFF file, and CFF entries are always of the type "Software".

FYI: as of version [`1.2.0`](https://github.com/citation-file-format/citation-file-format/releases/tag/1.2.0), CFF also supports datasets *experimentally* (`type: dataset`).

sdruskat commented:
> Similarly, CFF supports a list of 'identifiers' of different types (url, swh, other). JabRef does have support for these fields, but the CFF specs appear allow the user to add as many of these identifiers as they like (I'm not entirely sure about this, but it seems like it). So a single entry could have multiple SWH identifiers, which seems tricky to work into the JabRef system.

Yes, people can supply several identifiers for a software (which reflects reality).
Perhaps you could take a hierarchical approach, checking for some things first, then others, etc., and pick one entry based on a heuristic? E.g., 
1. CFF has root level `doi` field or exactly one `doi` entry in the `identifiers` section: take that
2. CFF has no DOIs, but has exactly one `swh` entry in the identifiers: take that
3. CFF has more than one `swh`-type identifier, including one with the [`rel` type](https://github.com/citation-file-format/citation-file-format/releases/tag/1.2.0): take that
4. etc.

Happy to help with such a heuristic which I think should ideally be based on the [software citation principles](https://peerj.com/articles/cs-86/) as much as possible.

> However, there are also some 'static' fields in CFF that don't have JabRef counterparts (specifically: commit, license-url, repository-code, repository-artifact). You mentioned earlier that if CFF fields don't map to JabRef standard fields, I could "import/add them with a prefix". I wasn't sure what that meant, and was unable to find any examples of it in the codebase, so I was wondering if I could get a little more clarification on that.

In terms of what the [software citation principles](https://peerj.com/articles/cs-86/) suggest, `repository-code` will contain a value that's potentially more important to keep than the others, and may even be used to override a generic `url` field. But that's semantics which I'm not sure you want to have in JabRef?

AidanM11 commented:
Alright, so I've changed the author handling to incorporate the JabRef Author system, as @Siedlerchr suggested, and it seems to be working well.
For the identifiers, I've implemented something similar to the heuristic @sdruskat suggested. As JabRef has separate fields for DOIs and SWHIDs, they have separate heuristics. Currently, they are:

For DOI:
1. If there is a root-level DOI, use that
2. If there is a single DOI in the identifiers list, use that
3. Otherwise, do not use any DOI.

For SWHID
1. If there is a single SWHID, use that.
2. If there are multiple SWHIDs, but only one with the "rel" tag, use that one.
3. Otherwise, do not use any SWHID.

As you can see, when there are multiple identifiers of the same type, the current system just discards them. This doesn't seem ideal, but as the CFF file does not contain any other information that could differentiate those identifiers, it seems like the only alternative is to choose a random one, and that seems like a bad option. I guess it could also put them all into the field in a comma-separated list, but that seems like it goes against the intended use of those fields.

Anyway, besides that, I've still got a small list of things to do before this pull request is finished (updated tests, handling fields without mappings, dataset type support, etc) so I'll keep working on it. Just wanted to make sure this identifier handling is satisfactory.

Siedlerchr commented:
I think this sounds good for the heuristic, DOI is always a good/valid option. Provides the easiest way to get any addtional metadata if needed. 
I also added a devcall label, so that we @JabRef/developers can discuss this in detail in the next devcall if we come up with another idea or addtional considerations.  But your approach so far sounds good

jspaaks commented:
Thanks @AidanM11 for doing this work! I like your heuristics. Perhaps one thing to consider is that it's relatively common to have two DOIs in `identifiers`: one for the version of the software, and one for the concept / collection / all versions. For such cases, neither would make it through if there isn't a root `doi`, right? I guess that is by design, since it is a bit of an undecidable problem. Anyway, an alternative could be to pick the first one and use that, something like `identifiers.filter(type=doi)[0]`. I understand though if you don't like to make this arbitrary choice. I've made a [suggestion](https://github.com/citation-file-format/citation-file-format/issues/69#issuecomment-551066432) over at Citation File Format to introduce some kind of qualifier like datacite's `isVersionOf`, `isCitedBy`, etc in order to help tooling decide which element of `identifiers` is most appropriate for their purposes.

Siedlerchr commented on src/test/java/org/jabref/logic/importer/fileformat/CffImporterTest.java:
```diff
@@ -83,4 +85,122 @@ public void testImportEntries() throws IOException, URISyntaxException {
+        List<BibEntry> bibEntries = importer.importDatabase(file, StandardCharsets.UTF_8).getDatabase().getEntries();
+        BibEntry entry = bibEntries.get(0);
+
+        assertEquals(entry.getField(StandardField.AUTHOR), Optional.of("Joe van Smith and Bob Jones, Jr."));
```
Just a litte bit more simpler solution. So you don't need an assertEquals in every line
Create a new BibEntry expected with the fields  and then compare it to the response
BibEntry expected = new BibEntry(...).withField(....)
Collection.singletonList(expected, bibEntries)

Siedlerchr commented:
Thanks for the work, can you please merge in the latest upstream main branch of JabRef and resolve the conflicts?

Siedlerchr commented on src/main/java/org/jabref/logic/importer/fileformat/CffImporter.java:
```diff
@@ -0,0 +1,213 @@
+
+        // Select DOI to keep
+        if (entryMap.get(StandardField.DOI) == null && citation.ids != null) {
+            List<CffIdentifier> doiIds = IntStream.range(0, citation.ids.size())
```
You don't need an IntStream here, you can for the DOI directly use   
```

Optional<CffIdentifier> cffDoi = citation.ids.stream().filter(id -> "doi".equals(id.type)).findFirst();

            cffDoi.ifPresent(doi-> {
                entryMap.put(StandardField.DOI, doi.value);
            });
```

Siedlerchr commented on src/main/java/org/jabref/logic/importer/fileformat/CffImporter.java:
```diff
@@ -0,0 +1,213 @@
+
+        // Select SWHID to keep
+        if (citation.ids != null) {
+            List<String> swhIds = IntStream.range(0, citation.ids.size())
```
Similar approach as above
       
`    swhIds = citation.ids.stream().filter(id -> "swh".equals(id.type)).map(id->id.value).collect(Collectors.toList());`

Siedlerchr commented on src/main/java/org/jabref/logic/importer/fileformat/CffImporter.java:
```diff
@@ -0,0 +1,213 @@
+                List<String> relSwhIds = swhIds.stream()
+                                               .filter(id -> id.split(":").length > 3) // quick filter for invalid swhids
+                                               .filter(id -> id.split(":")[2].equals("rel"))
+                                               .collect(Collectors.toList());
```
and here you can also use findFirst again as above

Siedlerchr commented on src/main/java/org/jabref/logic/importer/fileformat/CffImporter.java:
```diff
@@ -0,0 +1,213 @@
+            return false;
+        }
+
+        if (citation != null && citation.vals.get("title") != null) {
```
I would move this within the try block and simplify this as a one liner:
return citation!=null & citation.vals.get("title) != null

Siedlerchr commented on src/main/java/org/jabref/logic/importer/fileformat/CffImporter.java:
```diff
@@ -0,0 +1,213 @@
+    }
+
+    private HashMap<String, StandardField> getFieldMappings() {
+        HashMap<String, StandardField> hm = new HashMap<String, StandardField>();
```
I would propose a better name: e.g. mappedFields for the HashMap. makes it easier for others to understand it

Siedlerchr reviewed (changes requested):
Overall looks good to me and thanks for the tests! I have just some minor improvements regarding the code style.

AidanM11 commented:
@Siedlerchr I've made the requested changes. Thanks for the review!

calixtus reviewed (approved):
Fixed some minor style issues. Otherwise LGTM, thanks for making JabRef better.
As soon as tests are green, we're merging.
